dexpv2/segmentation.py

```diff
@@ -10,6 +11,135 @@
 LOG = logging.getLogger(__name__)
 LOG.setLevel(logging.INFO)
 
+try:
+    import cupy as xp
+
+    LOG.info("cupy found.")
+except (ModuleNotFoundError, ImportError):
+    import numpy as xp
+
+    LOG.info("cupy not found using numpy.")
+
+
+def discretize_multiple_float16_to_uint16(
+    float16_arrays: List["xp.ndarray"],  # type: ignore # 'xp' will be defined at runtime
+) -> Tuple[List["xp.ndarray"], "xp.ndarray"]:  # type: ignore
+    """
+    Discretizes multiple arrays (e.g., CuPy or NumPy) of float16 values to uint16,
+    preserving order using a global mapping across all arrays.
+
+    Args:
+        float16_arrays (List[xp.ndarray]): A list of arrays (e.g., CuPy or NumPy),
+                                          each with dtype float16.
+                                          'xp' should be the array module (e.g., numpy or cupy).
+
+    Returns:
+        tuple: A tuple containing:
+            - y_uint16_list (List[xp.ndarray]): A list of discretized arrays,
+                                               each with dtype uint16, corresponding
+                                               to the input arrays.
+            - uint16_to_float16_lookup (xp.ndarray): A single lookup table
+              (array of float16) for all arrays, where the index
+              is the uint16 value and the value is the corresponding
+              original float16 value.
+
+    Raises:
+        TypeError: If any input array's dtype is not float16.
+        ValueError: If the list of arrays is empty, or if the total number
+                    of unique values across all arrays exceeds the capacity
+                    of uint16 (65536).
+    """
+    # Ensure xp is defined (this is more of a runtime check if not using static analysis)
+    if "xp" not in globals() and "xp" not in locals():
+        raise NameError(
+            "Array library 'xp' is not defined. Please import numpy as xp or cupy as xp."
+        )
+
+    if not float16_arrays:
+        raise ValueError("Input list of arrays cannot be empty.")
+
+    # Validate input types and collect original shapes and sizes
+    original_shapes = []
+    original_sizes = []
+    for i, arr in enumerate(float16_arrays):
+        # Assuming 'xp' is defined, xp.ndarray would be the type to check against
+        # For simplicity and following user's snippet, primarily checking dtype.
+        if not hasattr(arr, "dtype") or arr.dtype != xp.float16:
```

Comment:
When would it not have `dype`? Because it's an array, I thought it must have `dtype`.